### PR TITLE
[FLINK-12617] [container] StandaloneJobClusterEntrypoint defaults to random JobID for non-HA setups

### PR DESCRIPTION
**Pull request by knaufk**
## Brief change log

* StandaloneJobClusterEntrypoint defaults to random JobID for non-HA setups
* refactor StandaloneJobClusterConfigurationParserFactoryTest to use temporary folder for global Flink configuration

## Verifying this change
  - run unit tests in flink-container to check the correct defaults are 
  - build Flink locally, build job-cluster image
  -  run a job without HA and check that JobID is random
```
./build.sh --job-jar ../../flink-examples/flink-examples-streaming/target/flink-examples-streaming_2.11-1.9-SNAPSHOT-TopSpeedWindowing.jar --from-local-dist
FLINK_JOB=org.apache.flink.streaming.examples.windowing.TopSpeedWindowing docker-compose up
```
  - run a job without HA and explicit JobID and check that parameter takes precedence
```
FLINK_JOB_ARGUMENTS="--job-id 00000000000000000000000000000000" FLINK_JOB=org.apache.flink.streaming.examples.windowing.TopSpeedWindowing docker-compose up
```

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`:  no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactoryTest.java [hunk 1]
```diff
@@ -47,9 +54,21 @@
...
+	public TemporaryFolder tempFolder = new TemporaryFolder();
+	private File confFile;
+	private String confDirPath;
+
+	@Before
+	public void createEmptyFlinkConfiguration() throws IOException {
+		File confDir = tempFolder.getRoot();
+		confDirPath = confDir.getAbsolutePath();
+		confFile = new File(confDir, GlobalConfiguration.FLINK_CONF_FILENAME);
+		new FileOutputStream(confFile).close();
```
Comment (tillrohrmann) [inline flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactoryTest.java]: I would suggest to replace with `confFile.createNewFile();`

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java [hunk 2]
```diff
@@ -105,10 +109,14 @@ private int getRestPort(CommandLine commandLine) throws FlinkParseException {
 		}
 	}
 
-	private static JobID getJobId(CommandLine commandLine) throws FlinkParseException {
+	private static JobID getJobId(CommandLine commandLine, Configuration configuration) throws FlinkParseException {
 		String jobId = commandLine.getOptionValue(JOB_ID_OPTION.getOpt());
 		if (jobId == null) {
-			return DEFAULT_JOB_ID;
+			if (configuration.getString(HighAvailabilityOptions.HA_MODE, null) == null) {
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: Better to replace with `!HighAvailabilityMode.isHighAvailabilityModeActivated(configuration)`

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java [hunk 3]
```diff
@@ -82,7 +86,7 @@ public StandaloneJobClusterConfiguration createResult(@Nonnull CommandLine comma
 		final int restPort = getRestPort(commandLine);
 		final String hostname = commandLine.getOptionValue(HOST_OPTION.getOpt());
 		final SavepointRestoreSettings savepointRestoreSettings = CliFrontendParser.createSavepointRestoreSettings(commandLine);
-		final JobID jobId = getJobId(commandLine);
+		final JobID jobId = getJobId(commandLine, getGlobalConfiguration(configDir, dynamicProperties));
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: I think we should move the decision which `JobID` to use out of the `StandaloneJobClusterConfigurationParserFactory`. That way we don't have to load the global configuration here which actually should only happen after the `StandaloneJobClusterConfiguration` has been parsed giving the configuration directory. The decision could be moved to `StandaloneJobClusterEntryPoint` by introducing a `createJobIdForCluster(clusterConfiguration.getOptionalJobId(), globalConfiguration)`

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactoryTest.java [hunk 4]
```diff
@@ -73,17 +96,37 @@ public void testEntrypointClusterConfigurationParsing() throws FlinkParseExcepti
...
+		assertThat(clusterConfiguration.getJobId(), is(not(equalTo(ZERO_JOB_ID))));
+	}
+
+	@Test
+	public void testJobIDDefaultsToZeroJobIdWithHA() throws FlinkParseException, FileNotFoundException {
+		final String[] args = {"--configDir", confDirPath};
+
+		final StandaloneJobClusterConfiguration clusterConfiguration = commandLineParser.parse(args);
+
+		configureHighAvailabilityGlobally();
```
Comment (tillrohrmann) [inline flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactoryTest.java]: this should be the first statement in the test

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactoryTest.java [hunk 5]
```diff
@@ -73,17 +96,37 @@ public void testEntrypointClusterConfigurationParsing() throws FlinkParseExcepti
...
+
+		assertThat(clusterConfiguration.getJobId(), is(not(equalTo(ZERO_JOB_ID))));
```
Comment (tillrohrmann) [inline flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactoryTest.java]: the `not` should not be here

**Comment by knaufk** [comment]
Thank you for the review @tillrohrmann. I thought about this as well and in the end chose to do it in the `ParserResultFactory`. I will update the PR accordingly.

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactoryTest.java (hunk 1 above)
Comment (knaufk) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactoryTest.java]: Done

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java (hunk 2 above)
Comment (knaufk) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: Done

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java (hunk 3 above)
Comment (knaufk) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: Done

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactoryTest.java (hunk 4 above)
Comment (knaufk) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactoryTest.java]: Douh, but test does not exist anymore in new version.

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactoryTest.java (hunk 5 above)
Comment (knaufk) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactoryTest.java]: Douh, but test does not exist anymore in new version.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java [hunk 6]
```diff
@@ -38,10 +39,9 @@
 	private final SavepointRestoreSettings savepointRestoreSettings;
 
 	@Nonnull
-	private final JobID jobId;
+	private final Optional<JobID> jobId;
 
-	@Nullable
-	private final String jobClassName;
+	private final Optional<String> jobClassName;
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java]: Same here.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java [hunk 7]
```diff
@@ -52,17 +56,16 @@
...
-	private final String jobClassName;
+	private final Optional<String> jobClassName;
 
 	private StandaloneJobClusterEntryPoint(
 			Configuration configuration,
 			@Nonnull JobID jobId,
 			@Nonnull SavepointRestoreSettings savepointRestoreSettings,
 			@Nonnull String[] programArguments,
-			@Nullable String jobClassName) {
+			Optional<String> jobClassName) {
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java]: I would like to revert this change.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java [hunk 8]
```diff
@@ -97,14 +100,27 @@ public static void main(String[] args) {
...
+	static JobID createJobIdForCluster(Optional<JobID> jobID, Configuration globalConfiguration) {
+		if (jobID.isPresent()) {
+			return jobID.get();
+		}
+
+		if (!HighAvailabilityMode.isHighAvailabilityModeActivated(globalConfiguration)) {
+			return JobID.generate();
+		} else {
+			return ZERO_JOB_ID;
+		}
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java]:
```
return jobID.orElseGet(
			() -> {
				if (HighAvailabilityMode.isHighAvailabilityModeActivated(globalConfiguration)) {
					return ZERO_JOB_ID;
				} else {
					return JobID.generate();
				}
			});
```

Seems a bit more idiomatic.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java [hunk 9]
```diff
@@ -63,8 +63,7 @@
 	@Nonnull
 	private final String[] programArguments;
 
-	@Nullable
-	private final String jobClassName;
+	private final Optional<String> jobClassName;
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: These changes should be unrelated if I'm not mistaken. Please revert.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java [hunk 10]
```diff
@@ -105,13 +104,13 @@ private int getRestPort(CommandLine commandLine) throws FlinkParseException {
 		}
 	}
 
-	private static JobID getJobId(CommandLine commandLine) throws FlinkParseException {
+	private static Optional<JobID> getJobId(CommandLine commandLine) throws FlinkParseException {
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: I think it would also be ok to simply return a nullable `JobID` since it's an internal method.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java [hunk 11]
```diff
@@ -97,14 +100,27 @@ public static void main(String[] args) {
 
 		StandaloneJobClusterEntryPoint entrypoint = new StandaloneJobClusterEntryPoint(
 			configuration,
-			clusterConfiguration.getJobId(),
+			createJobIdForCluster(clusterConfiguration.getJobId(), configuration),
 			clusterConfiguration.getSavepointRestoreSettings(),
 			clusterConfiguration.getArgs(),
 			clusterConfiguration.getJobClassName());
```
Comment (tillrohrmann) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java]: If `getJobClassName` returns an `Optional`, then we should insert `clusterConfiguration.getJobClassName().orElse(null)`

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPointTest.java [hunk 12]
```diff
@@ -59,6 +66,60 @@ public void testDontOverwriteExecutionMode() {
...
 
+	@Test
+	public void configuredJobIDTakesPrecedenceWithHA() {
+		Optional<JobID> jobID = Optional.of(JobID.generate());
+
+		Configuration globalConfiguration = new Configuration();
+		globalConfiguration.setString(HighAvailabilityOptions.HA_MODE, "zookeeper");
+
+		JobID jobIdForCluster = StandaloneJobClusterEntryPoint.createJobIdForCluster(
+				jobID,
```
Comment (tillrohrmann) [inline flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPointTest.java]: Wrong indentation.

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPointTest.java [hunk 13]
```diff
@@ -59,6 +66,60 @@ public void testDontOverwriteExecutionMode() {
...
+
+		assertThat(jobIdForCluster, is(jobID.get()));
+	}
+
+	@Test
+	public void jobIDdefaultsToZeroWithHA() {
+		Optional<JobID> jobID = Optional.empty();
+
+		Configuration globalConfiguration = new Configuration();
+		globalConfiguration.setString(HighAvailabilityOptions.HA_MODE, "zookeeper");
```
Comment (tillrohrmann) [inline flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPointTest.java]: one could encapsulate this call into `enableHighAvailability(Configuration)` and that way reduce duplication.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java [hunk 14]
```diff
@@ -121,8 +120,8 @@ private PackagedProgram createPackagedProgram() throws FlinkException {
 	}
 
 	private String getJobClassNameOrScanClassPath() throws FlinkException {
-		if (jobClassName != null) {
-			return jobClassName;
+		if (jobClassName.isPresent()) {
```
[the remark this answers is not in the document]
Comment (knaufk) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: Done.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java [hunk 15]
```diff
@@ -38,10 +39,9 @@
 	private final SavepointRestoreSettings savepointRestoreSettings;
 
 	@Nonnull
-	private final JobID jobId;
+	private final Optional<JobID> jobId;
```
[the remark this answers is not in the document]
Comment (knaufk) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java]: Done.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java (hunk 6 above)
Comment (knaufk) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java]: Done.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java (hunk 7 above)
Comment (knaufk) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java]: Done.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java (hunk 8 above)
Comment (knaufk) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java]: I agree.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java (hunk 10 above)
Comment (knaufk) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfigurationParserFactory.java]: Done

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java (hunk 11 above)
Comment (knaufk) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java]: Its not returning an Optional anymore.

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPointTest.java (hunk 12 above)
Comment (knaufk) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPointTest.java]: Hopefully, done. Not completely sure, what the correct indentation is here, because my Flikn checkstyle profile actually produces this.

File: flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPointTest.java (hunk 13 above)
Comment (knaufk) [reply flink-container/src/test/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPointTest.java]: Done

**Comment by knaufk** [comment]
@tillrohrmann thanks for the review. I reverted the change to `Optional` for fields and addressed the other comments.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java (hunk 9 above)
Comment (knaufk) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java]: Done. I originally did these changes, because I did not want to mix `@Nullable` xor `Optional` for fields.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java [hunk 16]
```diff
@@ -54,7 +54,7 @@
 			@Nullable String jobClassName) {
 		super(configDir, dynamicProperties, args, hostname, restPort);
 		this.savepointRestoreSettings = requireNonNull(savepointRestoreSettings, "savepointRestoreSettings");
-		this.jobId = requireNonNull(jobId, "jobId");
```
Comment (1u0) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java]: The annotation `@Nonnull JobID jobId` in the constructor parameter mismatches with new constrain.

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java [hunk 17]
```diff
@@ -97,14 +102,26 @@ public static void main(String[] args) {
 
 		StandaloneJobClusterEntryPoint entrypoint = new StandaloneJobClusterEntryPoint(
 			configuration,
-			clusterConfiguration.getJobId(),
+			createJobIdForCluster(Optional.ofNullable(clusterConfiguration.getJobId()), configuration),
```
Comment (1u0) [inline flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java]: Personally, I'd change it to:
```
Optional.ofNullable(clusterConfiguration.getJobId())
    .orElseGet(() -> createJobIdForCluster(configuration))
```

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java (hunk 16 above)
Comment (knaufk) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterConfiguration.java]: Done

File: flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java (hunk 17 above)
Comment (tillrohrmann) [reply flink-container/src/main/java/org/apache/flink/container/entrypoint/StandaloneJobClusterEntryPoint.java]: I think this is a good improvement because it simplifies `createJobIdForCluster`

**Comment by tillrohrmann** [comment]
Updating this PR with suggestions from @1u0 and force pushing. So don't be surprised.
